<html>

<head>
    <title>Legacy euk annotation database mapping to BSML</title>
    <link rel="stylesheet" href="documentation.css">
</head>

<body>

<p class="doc_header">
    Legacy euk annotation database mapping to BSML<br>
</p>

This document provides a mapping between fields in a legacy eukaryotic
annotation database and XML elements in BSML.  The usage conventions in BSML
have been tailored to match usage/encoding conventions in Chado.  For example,
the gene encoding in BSML mirrors the standard Chado gene encoding.  Additional
information about Chado usage conventions can be found <a href=''>here</a>.

<p class="section_header">
    Genomes element
</p>

<p class="subsection_header">
    Organism and Cross-reference
</p>

<p class="xpath">
    //Genome/Organism/@genus = the first word in common..genomes.name
    //Genome/Organism/@species = everything after the first word in common..genomes.name

    //Genome/Organism/Attribute/[@name='abbreviation'] = common..genomes.file_moniker
    //Genome/Organism/Attribute/[@name='gram_strain'] = new_project.gram_stain
    //Genome/Organism/Attribute/[@name='genetic_code'] = new_project.genetic_code

    //Genome/Cross-reference/@database = 'TIGR_euk:' + legacy database name
    //Genome/Cross-reference/@identifier = common..genomes.file_moniker
    //Genome/Cross-reference/@identifier_type = 'current'
</p>


<p class="section_header">
    Sequences element
</p>

<p class="subsection_header">
    Naming convention
</p>
<p>
    The identifier for the sequence is generated according to the <a
    href='http://tools/viewcvs/viewcvs.cgi/*checkout*/ANNOTATION/chado/doc/naming_conventions_for_features.html?rev=HEAD'>naming conventions document </a>.  The placeholder '_id_' will be used in place of an actual id.
</p>
<p class="xpath">
    //Sequence/@id = lma1.gene.1
</p>
<p>
    Example:<br>
    lma1.gene.1
</p>
<p>
    Note: All subfeatures of the sequence will use a similar naming convention to generate 
    <xpath>//Feature/@id</xpath>. The naming convention for <xpath>//Sequence/@id</xpath> and 
    <xpath>//Feature/@id</xpath> will likely change.  Software should not make any assumptions 
    about the naming conventions.
</p>
<p class="subsection_header">
    Assembly sequence
</p>
<p class="xpath">
    //Sequence/@id = _id_
    //Sequence/@class = `assembly'
    //Sequence/@molecule = `dna'
    //Sequence/@length = datalength(assembly.sequence) 
    //Sequence/@topology = 'linear'

    //Sequence/Attribute[@name='molecule_name'] = clone_info.clone_name

    //Sequence/Attribute[@name='molecule_type'] = 



    //Sequence/Attribute-list/Attribute[@name='SO']/@value = `supercontig' 
    ??
    //Sequence/Attribute-list/Attribute[@name='SO']/@value = `chromosome' 
    ??

    //Sequence/Seq-data->text() = assembly.sequence
</p>
<p class="subsection_header">
    Protein sequence
</p>
<p class="xpath">
    //Sequence/@id = _id_."_seq"
    //Sequence/@class = `protein'
    //Sequence/@molecule = `aa'
    //Sequence/@length = datalength(asm_feature.protein) 
        where asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
    //Sequence/Seq-data/@identifier = _id_
    //Sequence/Seq-data->text() = asm_feature.sequence 
        where asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
</p>
<p class="subsection_header">
    CDS sequence
</p>
<p>
    The ORF sequence is stored as the CDS sequence.  This encoding is consistent with the 
    Chado gene encoding.
</p>
<p class="xpath">
    //Sequence/@id = _id_."_seq"
    //Sequence/@class = `CDS'
    //Sequence/@molecule = `dna'
    //Sequence/@length = datalength(asm_feature.sequence) 
        where asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
    //Sequence/Seq-data->text() = asm_feature.sequence 
        where asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
</p>
<p class="subsection_header">
    tRNA sequence
</p>
<p class="xpath">
    //Sequence/@id = _id_
    //Sequence/@class = `tRNA'
    //Sequence/@molecule = `dna'
    //Sequence/@length = datalength(asm_feature.sequence) 
        where asm_feature.feat_type = `tRNA'
    //Sequence/Seq-data->text() = asm_feature.sequence 
        where feat_type = `tRNA'
</p>
<p class="subsection_header">
    rRNA sequence
</p>
<p class="xpath">
    //Sequence/@id = _id_
    //Sequence/@class = `rRNA'
    //Sequence/@molecule = `dna'
    //Sequence/@length = datalength(asm_feature.sequence) 
        where asm_feature.feat_type = `rRNA'
    //Sequence/Seq-data->text() = asm_feature.sequence 
        where feat_type = `rRNA'
</p>
<p class="subsection_header">
    snRNA sequence
</p>
<p class="xpath">
??
</p>
<p>
    The following sections describe <Feature> elements which are nested below the assembly 
    sequence described above.
</p>

<p class="section_header">
    Feature element
</p>
<p>
    The TU, model, and exon features from the legacy annotation database are
    used to fill features in BSML that correspond to the standard Chado gene
    encoding.  The Chado gene encoding requires 5 &lt;Feature&gt; 
    elements: gene, transcript, CDS, exon, and protein.

    Alternative splicing is handled by associating the features of each isoform
    with a single gene feature.  Additional information about the Chado
    encoding for alternative splicing can be found in this <a
    href=''>powerpoint presentation</a>.

    This mapping only pertains to TU, which have at least one model of type "working".
</p>
<p class="subsection_header">
    Gene Feature
</p>
<p class="xpath">
    //Feature[@class=`gene`]/@id = _id_
    //Feature[@class=`gene`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 < asm_feature.end3
        and asm_feature.feat_type = `TU'
    //Feature[@class=`gene`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 < asm_feature.end3
        and asm_feature.feat_type = `TU'
    //Feature[@class=`gene`]/Interval-loc/complement = 0
    //Feature[@class=`gene`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 > asm_feature.end3
        and asm_feature.feat_type = `TU'
    //Feature[@class=`gene`]/Interval-loc/endpos = asm_feature.end3 
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`gene`]/Interval-loc/complement = 1
</p>
<p class="subsection_header">
    Gene Database xrefs
</p>
<p class="xpath">
    //Feature[@class=`gene`]/Cross-reference/@database='...'
    //Feature[@class=`gene`]/Cross-reference/@identifier=asm_feature.feat_name
    //Feature[@class=`gene`]/Cross-reference/@identifier-type='feat_name'

    //Feature[@class=`gene`]/Cross-reference/@database='...'
    //Feature[@class=`gene`]/Cross-reference/@identifier=ident.locus
	   where ident.feat_name = asm_feature.feat_name
    //Cross-reference/@identifier-type='locus'
    //Feature[@class=`gene`]/Cross-reference/@identifier=ident.pub_locus
	   where ident.feat_name = asm_feature.feat_name
    //Cross-reference/@identifier-type='display_locus'
    
    //Feature-group[@group-set='gene']/Feature-group-member/@feature-type='gene'
    //Feature-group[@group-set='gene']/Feature-group-member/@featref= _id_

</p>
<p class="subsection_header">
    Transcript Feature
</p>
<p class="xpath">
    //Feature[@class=`transcript`]/@id = _id_
    //Feature[@class=`transcript`]/Interval-loc/startpos = asm_feature.end5 - 1 
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript`]/Interval-loc/endpos = asm_feature.end3 
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = 'TU'
    //Feature[@class=`transcript`]/Interval-loc/complement = 0
    //Feature[@class=`transcript`]/Interval-loc/startpos = asm_feature.end5 - 1 
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript`]/Interval-loc/endpos = asm_feature.end3 
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript`]/Interval-loc/complement = 1
</p>
<p class="subsection_header">
    Transcript Database xrefs
</p>
<p class="xpath">
    //Feature[@class=`transcript`]/Cross-reference/@database='...'
    //Feature[@class=`transcript`]/Cross-reference/@identifier=asm_feature.feat_name
    //Feature[@class=`transcript`]/Cross-reference/@identifier-type='feat_name'

    //Feature[@class=`transcript`]/Cross-reference/@database='...'
    //Feature[@class=`transcript`]/Cross-reference/@identifier=ident.locus
        where ident.feat_name = asm_feature.feat_name
	and asm_feature.feat_type = 'TU'
    //Feature[@class=`transcript`]/Cross-reference/@identifier-type='locus'

    //Feature[@class=`transcript`]/Cross-reference/@database='...'
    //Feature[@class=`transcript`]/Cross-reference/@identifier=ident.alt_locus
        where ident.feat_name = asm_feature.feat_name
	and asm_feature.feat_type = 'TU'
    //Feature[@class=`transcript`]/Cross-reference/@identifier-type='alt_locus'

    //Feature[@class=`transcript`]/Cross-reference/@database='...'
    //Feature[@class=`transcript`]/Cross-reference/@identifier=ident.pub_locus
        where ident.feat_name = asm_feature.feat_name
	and asm_feature.feat_type = 'TU'
    //Feature[@class=`transcript`]/Cross-reference/@identifier-type='pub_locus'

    //Feature-group[@group-set='gene']/Feature-group-member/@feature-type=transcript'
    //Feature-group[@group-set='gene']/Feature-group-member/@featref= _id_
</p>
<p class="subsection_header">
    CDS Feature
</p>
<p class="xpath">
    //Feature[@class='CDS`]/@id = _id_
    //Feature[@class=`CDS`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
    //Feature[@class=`CDS`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 < asm_feature.end3
        and asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
    //Feature[@class=`CDS`]/Interval-loc/complement = 0
    //Feature[@class=`CDS`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 > asm_feature.end3
        and asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
    //Feature[@class=`CDS`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 > asm_feature.end3
        and asm_feature.feat_type = `model'
	and asm_feature.feat_name = phys_ev.feat_name
	and phys_ev.ev_type = 'working'
    //Feature[@class=`CDS`]/Interval-loc/complement = 1
</p>
<p class="subsection_header">
    CDS Database xrefs
</p>
<p class="xpath">
    //Feature[@class=`CDS`]/Cross-reference/@database='...'
    //Feature[@class=`CDS`]/Cross-reference/@identifier=asm_feature.feat_name
    //Feature[@class=`CDS`]/Cross-reference/@identifier-type='feat_name'

    //Feature[@class=`transcript`]/Cross-reference/@database='...'
    //Feature[@class=`transcript`]/Cross-reference/@identifier=ident.locus
        where ident.feat_name = asm_feature.feat_name
	and asm_feature.feat_type = 'TU'

    //Feature-group[@group-set='gene']/Feature-group-member/@feature-type='CDS'
    //Feature-group[@group-set='gene']/Feature-group-member/@featref= _id_
</p>
<p class="subsection_header">
    CDS Sequence residues
</p>
<p class="xpath">
    //Feature[@class=`CDS']/Link[@rel='sequence']/@href = _id_+"_seq"

    Example:
    
    <incomplete>[XML snippet here]<incomplete>
</p>
<p class="subsection_header">
    Exon Feature
</p>
<p class="xpath">
    //Feature[@class='exon`]/@id = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_exon"
    //Feature[@class=`exon`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 < asm_feature.end3
        and asm_feature.feat_type = `exon'
    //Feature[@class=`exon`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 < asm_feature.end3
        and asm_feature.feat_type = `exon'
    //Feature[@class=`exon`]/Interval-loc/complement = 0
    //Feature[@class=`exon`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 > asm_feature.end3
        and asm_feature.feat_type = `exon'
    //Feature[@class=`exon`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 > asm_feature.end3
        and asm_feature.feat_type = `exon'
    //Feature[@class=`exon`]/Interval-loc/complement = 1

    //Feature-group[@group-set='gene']/Feature-group-member/@feature-type='exon'
    //Feature-group[@group-set='gene']/Feature-group-member/@featref= database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_exon"
</p>
<p class="subsection_header">
    Exon Database xrefs
</p>
<p class="xpath">
    //Feature[@class='exon`]/Cross-reference/@database='...'
    //Feature[@class=`exon`]/Cross-reference/@identifier=asm_feature.feat_name
    where feat_type = 'exon'
    //Feature[@class=`exon`]/Cross-reference/@identifier-type='feat_name'

    Example:
    <incomplete>[XML snippet here]</incomplete>
</p>
<p class="subsection_header">
    Protein Features
</p>
<p class="xpath">
    //Feature[@class='protein`]/@id = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_protein"
    //Feature[@class=`protein`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `model'
    //Feature[@class=`protein`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `model'
    //Feature[@class=`protein`]/Interval-loc/complement = 0
    //Feature[@class=`protein`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `model'
    //Feature[@class=`protein`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `model'
    //Feature[@class=`protein`]/Interval-loc/complement = 1

    //Feature[@class=`protein']/Link[@rel='sequence']/@href = database_name
        +assembly.asmbl_id
        +"_"+asm_feature.feat_name where feat_type = `ORF'
        + "_protein_seq"

    //Feature-group[@group-set='gene']/Feature-group-member/@feature-type='protein'
    //Feature-group[@group-set='gene']/Feature-group-member/@featref= database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_protein"

    Example:
    <incomplete>[XML snippet here]<incomplete>
</p>

<p class="section_header">
    Feature relationships
</p>
<p>
    <incomplete>add description here</incomplete>
</p>

<p class="section_header">
    RNA Features
</p>
<p class="subsection_header">
    tRNA Features
</p>
<p class="xpath">
    //Feature[@class='tRNA`]/@id = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_tRNA"
    //Feature[@class=`tRNA`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `tRNA'
    //Feature[@class=`tRNA`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `tRNA'
    //Feature[@class=`tRNA`]/Interval-loc/complement = 0
    //Feature[@class=`tRNA`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `tRNA'
    //Feature[@class=`tRNA`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `tRNA'
    //Feature[@class=`tRNA`]/Interval-loc/complement = 1
</p>
<p class="subsection_header">
    tRNA Database xrefs
</p>
<p class="xpath">
    //Feature[@class=`tRNA`]/Cross-reference/@database='...'
    //Feature[@class=`tRNA`]/Cross-reference/@identifier=asm_feature.feat_name
    //Feature[@class=`tRNA`]/Cross-reference/@identifier-type='feat_name'
</p>
<p class="xpath">
    //Feature[@class=`tRNA`]/Cross-reference/@database='...'
    //Feature[@class=`tRNA`]/Cross-reference/@identifier=uppercase(legacy
    annotation database name) + ' ' + asm_feature.feat_name
    //Feature[@class=`tRNA`]/Cross-reference/@identifier-type='database_feat_name'
</p>
<p class="subsection_header">
    tRNA Sequence residues
</p>
<p class="xpath">
    //Feature[@class=`tRNA']/Link[@rel='sequence']/@href = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_tRNA_seq"
</p>
<p class="subsection_header">
    tRNA Annotation attributes
</p>
<p class="xpath">
    //Feature[@class='tRNA']/Attribute[@name='name'] = ident.com_name 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `tRNA'

    //Feature[@class='tRNA']/Attribute[@name='gene_sym'] = ident.gene_sym
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `tRNA'

    //Feature[@class='tRNA']/Attribute[@name='pub_comment'] = ident.pub_comment 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `tRNA'

    //Feature[@class='tRNA']/Attribute[@name='anti-codon'] = feat_score.score
        where asm_feature.feat_type = `tRNA'
	and asm_feature.feat_name = ORF_attribute.feat_name
	and ORF_attribute.att_type = 'tRNA'
	and ORF_attribute.id = feat_score.input_id
	and feat_score.score_id = score_type.id
	and score_type.input_type = 'tRNA'
	and score_type.score_type = 'anti-codon'

    Example:
    <incomplete>[XML snippet here]<incomplete>
</p>
<p class="subsection_header">
    rRNA Encoding
</p>
<p class="xpath">
    //Feature[@class='rRNA`]/@id = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_rRNA"
    //Feature[@class=`rRNA`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `rRNA'
    //Feature[@class=`rRNA`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `rRNA'
    //Feature[@class=`rRNA`]/Interval-loc/complement = 0
    //Feature[@class=`rRNA`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `rRNA'
    //Feature[@class=`rRNA`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `rRNA'
    //Feature[@class=`rRNA`]/Interval-loc/complement = 1
</p>
<p class="subsection_header">
    rRNA Database xrefs
</p>
<p class="xpath">
    //Feature[@class=`rRNA`]/Cross-reference/@database='...'
    //Feature[@class=`rRNA`]/Cross-reference/@identifier=asm_feature.feat_name
    //Feature[@class=`rRNA`]/Cross-reference/@identifier-type='feat_name'
</p>
<p class="xpath">
    //Feature[@class=`rRNA`]/Cross-reference/@database='...'
    //Feature[@class=`rRNA`]/Cross-reference/@identifier= uppercase(legacy annotation
    database name) + ' ' + asm_feature.feat_name
    //Feature[@class=`rRNA`]/Cross-reference/@identifier-type='database_feat_name'
</p>
<p class="subsection_header">
    rRNA Sequence residues
</p>
<p class="xpath">
    //Feature[@class=`rRNA']/Link[@rel='sequence']/@href = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_rRNA_seq"
</p>
<p class="subsection_header">
    rRNA Annotation attributes
</p>
<p class="xpath">
    //Feature[@class='rRNA']/Attribute[@name='name'] = ident.com_name 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `rRNA'

    //Feature[@class='rRNA']/Attribute[@name='gene_sym'] = ident.gene_sym
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `rRNA'

    //Feature[@class='rRNA']/Attribute[@name='pub_comment'] = ident.pub_comment 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `rRNA'

    Example:
    <incomplete>[XML snippet here]</incomplete>
</p>
<p class="subsection_header">
    snRNA Encoding
</p>
</p>
    The SO identifier to describe this feat_type is `ncRNA'.
</p>
<p class="xpath">
    //Feature[@class='snRNA`]/@id = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_snRNA"
    //Feature[@class=`snRNA`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `sRNA'
    //Feature[@class=`snRNA`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 < asm_feature.end3 
        and asm_feature.feat_type = `sRNA'
    //Feature[@class=`snRNA`]/Interval-loc/complement = 0
    //Feature[@class=`snRNA`]/Interval-loc/startpos = asm_feature.end5 - 1
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `sRNA'
    //Feature[@class=`snRNA`]/Interval-loc/endpos = asm_feature.end3
        where asm_feature.end5 > asm_feature.end3 
        and asm_feature.feat_type = `sRNA'
    //Feature[@class=`snRNA`]/Interval-loc/complement = 1
</p>
<p class="subsection_header">
    snRNA Database xrefs
</p>
<p class="xpath">
    //Feature[@class=`snRNA`]/Cross-reference/@database='...'
    //Feature[@class=`snRNA`]/Cross-reference/@identifier=asm_feature.feat_name
    //Feature[@class=`snRNA`]/Cross-reference/@identifier-type='feat_name'
</p>
<p class="xpath">
    //Feature[@class=`snRNA`]/Cross-reference/@database='...'
    //Feature[@class=`snRNA`]/Cross-reference/@identifier=uppercase(legacy
    annotation database name) + ' ' + asm_feature.feat_name
    //Feature[@class=`snRNA`]/Cross-reference/@identifier-type='database_feat_name'
</p>
<p class="xpath">
    //Feature[@class=`snRNA`]/Cross-reference/@database='...'
    //Feature[@class=`snRNA`]/Cross-reference/@identifier=uppercase(legacy
    annotation database name) + ' ' + asm_feature.feat_name
    //Feature[@class=`snRNA`]/Cross-reference/@identifier-type='database_feat_name'
</p>
<p class="subsection_header">
    snRNA Sequence residues
</p>
<p class="xpath">
    //Feature[@class=`snRNA']/Link[@rel='sequence']/@href = database
        +"_"
        +assembly.asmbl_id
        +"_"
        +asm_feature.feat_name
        +"_snRNA_seq"
</p>
<p class="subsection_header">
    snRNA Annotation attributes
</p>
<p class="xpath">
    //Feature[@class='ncRNA']/Attribute[@name='name'] = ident.com_name
    where ident.feat_name = asm_feature.feat_name 
    and asm_feature.feat_type = `sRNA'

    //Feature[@class='sRNA']/Attribute[@name='gene_sym'] = ident.gene_sym
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `sRNA'

    //Feature[@class='sRNA']/Attribute[@name='pub_comment'] = ident.pub_comment 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `sRNA'

    Example:
    <incomplete>[XML snippet here]</imcomplete>
</p>

<p class="section_header">
    Other features
</p>

<p class="subsection_header">
    Ident information
</p>
<p class="xpath">
    //Feature[@class=`transcript']/Attribute[name='gene_product_name']/@content = ident.com_name 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript']/Attribute[name='assignby']/@content = ident.assignby 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript']/Attribute[name='date']/@content = ident.date 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript']/Attribute[name='comment']/@content = ident.comment 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript']/Attribute[name='auto_comment']/@content = ident.auto_comment 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript']/Attribute[name='gene_symbol']/@content = ident.gene_sym 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `TU'
    //Feature[@class=`transcript']/Attribute-list/Attribute[@name='EC']/@content = ident.ec# 
        where ident.feat_name = asm_feature.feat_name 
        and asm_feature.feat_type = `TU'
    //Feature[@class='transcript']/Attribute-list/Attribute[@name='SO']/@content = 'pseudogene'
        where ident.feat_name = asm_feature.feat_name
	and asm_feature.feat_type = 'TU'
	and ident.is_pseudogene = 1

The order of the attributes determines the relative rank of the attributes.  Attributes pulled from ident_xref should be stored in BSML ordered by ident_xref.relrank

    //Feature[@class=`transcript']/Attribute-list/Attribute[@name='EC']/@content = ident_xref.ident_val
        where ident_xref.feat_name = asm_feature.feat_name
	and ident_xref.xref_type = 'ec number'
        and asm.feature.feat_type = 'TU'
    //Feature[@class=`transcript']/Attribute[name='gene_product_name']/@content = ident_xref.ident_val
        where ident_xref.feat_name = asm_feature.feat_name
	and ident_xref.xref_type = 'product name'
        and asm.feature.feat_type = 'TU'
    //Feature[@class=`transcript']/Attribute[name='gene_name']/@content = ident_xref.ident_val
        where ident_xref.feat_name = asm_feature.feat_name
	and ident_xref.xref_type = 'gene name'
        and asm.feature.feat_type = 'TU'
    //Feature[@class=`transcript']/Attribute[name='gene_symbol']/@content = ident_xref.ident_val
        where ident_xref.feat_name = asm_feature.feat_name
	and ident_xref.xref_type = 'gene symbol'
        and asm.feature.feat_type = 'TU'

    //Feature[@class=`transcript`]/Cross-reference/@database='genbank'
    //Feature[@class=`transcript`]/Cross-reference/@identifier=ident_xref.ident_val
        where ident_xref.feat_name = asm_feature.feat_name
	and ident_xref.xref_type = 'genbank accession'
    //Cross-reference/@identifier-type='genbank_pid'
</p>
<p>
    Example:<br>
    <incomplete>[XML snippet here]</incomplete>
</p>

<p class="subsection_header">
    TIGR Roles
</p>
<p>
    The elements described in this section are assumed to contain the following 
    XPath prefix: <xpath>//Feature[@class=`transcript']</xpath>.  It has been replaced with `?' 
    for brevity.
</p>
<p class="xpath">
    //Attribute-list/Attribute[@name='TIGR_role']/@content = role_link.role_id
        where role_link.feat_name = asm_feature.feat_name

    //Attribute-list/Attribute[@name='assignby']/@content = role_link.assignby
        where role_link.feat_name = asm_feature.feat_name
 
    //Attribute-list/Attribute[@name='date']/@content = role_link.date 
        where role_link.feat_name = asm_feature.feat_name

    //Attribute-list/Attribute[@name='comment']/@content = notes.role_note
        where role_link.feat_name = asm_feature.feat_name
	and role_line.role_id = note.role_id
</p>
<p>
    Example:<br>
    <incomplete>[XML snippet here]</incomplete>
</p>
<p class="subsection_header">
    GO
</p>
<p>
    The elements described in this section are assumed to contain the following 
    XPath prefix: //Feature[@class=`transcript'].  It has been replaced with `?' 
    for brevity.
</p>
<p class="xpath">
    ?/Attribute-list/Attribute/[@name='GO']/@content = go_role_link.go_id 
        where go_role_link.feat_name = asm_feature.feat_name

    ?/Attribute-list/Attribute[@name='assignby']/@content = go_role_link.assign_by 
        where role_link.feat_name = asm_feature.feat_name

    ?/Attribute-list/Attribute[@name='date']/@content = go_role_link.date 
        where role_link.feat_name = asm_feature.feat_name

    ?/Attribute-list/Attribute/[@qualifier='qualifier']/@content= go_role_link.qualifier 
        where go_role_link.feat_name = asm_feature.feat_name

    ?/Attribute-list/Attribute/@name = go_evidence.ev_code 
        where go_evidence.role_link_id = go_role_link.id

    ?/Attribute-list/Attribute/@content = go_evidence.evidence 
</p>
<p>
    If go_evidence.with_ev is NOT NULL, an additional string is concatenated to the 
    <xpath>//Attribute/@content</xpath> attribute defined above.
</p>
<p class="xpath">
    ?/Attribute-list/Attribute/@content += " WITH "+go_evidence.with_ev where go_evidence.role_link_id = go_role_link_id
</p>
<p>
    Example:<br>
    <incomplete>[XML snippet here]</incomplete>
</p>
<p class="subsection_header">
    ORF_attribute
</p>
<p>
    A subset of information stored in the ORF_attribute table is encoded in 
    <xpath>//Feature[@class=`transcript']/Attribute</xpath>.
</p>
<p class="xpath">
    //Feature[@class=`transcript']/Attribute/@name = `MW'
    //Feature[@class=`transcript']/Attribute/@content = feat_score.score 
        where ORF_attribute.att_type = `MW' 
        and ORF_attribute.id = feat_score.input_id 
        and feat_score.score_id = common..score_type.id 
        and common..score_type.input_type = MW' 
        and common..score_type.score_type = `MW'

    //Feature[@class=`transcript']/Attribute/@name = `pI'
    //Feature[@class=`transcript']/Attribute/@content = feat_score.score 
        where ORF_attribute.att_type = `PI' 
        and ORF_attribute.id = feat_score.input_id 
        and feat_score.score_id = common..score_type.id 
        and common..score_type.input_type = `PI' 
        and common..score_type.score_type = `PI'

    //Feature[@class=`transcript']/Attribute/@name = `percent_GC'
    //Feature[@class=`transcript']/Attribute/@content = feat_score.score 
        where ORF_attribute.att_type = `GC' 
        and ORF_attribute.id = feat_score.input_id 
        and feat_score.score_id = common..score_type.id 
        and common..score_type.input_type = `GC' 
        and common..score_type.score_type = `GC'

    //Feature[@class=`transcript']/Attribute/@name = `transmembrane_coords'
    //Feature[@class=`transcript']/Attribute/@content = feat_score.score 
        where ORF_attribute.att_type = `GES' 
        and ORF_attribute.id = feat_score.input_id 
        and feat_score.score_id = common..score_type.id 
        and common..score_type.input_type = `GES' 
        and common..score_type.score_type = `coords'

    //Feature[@class=`transcript']/Attribute/@name = `transmembrane_regions
    //Feature[@class=`transcript']/Attribute/@content = feat_score.score 
        where feat_score.input_id = ORF_attribute.id 
        and ORF_attribute.att_type = `GES' 
        and common..score_type.id = feat_score.score_id 
        and common..score_type = `regions'

    //Feature[@class=`transcript']/Attribute/@name = `lipo_membrane_protein'
    //Feature[@class=`transcript']/Attribute/@content = feat_score.score 
        where feat_score.input_id = ORF_attribute.id 
        and ORF_attribute.att_type = `LP'

    //Feature[@class=`transcript']/Attribute/@name = `outer_membrane_protein'
    //Feature[@class=`transcript']/Attribute/@content = feat_score.score 
        where feat_score.input_id = ORF_attribute.id 
        and ORF_attribute.att_type = `OMP'
</p>
<p>
    Example:<br>
    <incomplete>[XML snippet here]</incomplete>
</p>

<p class="subsection_header">
    Pseudomolecule mapping
</p>
<p>
    <incomplete>INCOMPLETE</incomplete>
</p>

<p class="section_header">
    Evidence
</p>
<p>
    Search evidence is encoded using &lt;Seq-pair-alignment&gt; and &lt;Seq-pair-run&gt; elements 
    in BSML.  &lt;Sequence&gt; elements are required for all referenced sequence identifiers.  
    These sparsely populated &lt;Sequence&gt; elements are referred to as sequence stubs in 
    this document.
</p>

<p class="section_header">
    Table Usage Summary
</p>

<p class="subsection_header">
    Migrated tables
</p>
<p>
    This information was parsed from the Sybase sysobjects table for a Prok database.
</p>
<p>
    <sql>select name from sysobjects where type = `U'</sql>
</p>
<p>
    The following tables where considered during the migration.  Fields that are currently 
    used in the migration to BSML are listed and categorized into (3) categories. 
    <ul>
        <li>
            Fields that are not required for the BSML migration are striked.   These fields are 
            ignored by the migration script and have been deemed unimportant for purposes of 
            the migration.
        </li>
        <li>
            Fields that are a primary source of data that is migrated to BSML are in bold.  All 
            of these fields should have a documented mapping to a field in BSML in the previous 
            sections of this document.
        </li>
        <li>
            Fields that are not migrated as part of the current implementation are marked with 
            a circle.  These fields are may be considered as part of future migration 
            development 
        </li>
        <li>
            All other fields (non-boldface, non-striked, no asterisk) are either required for 
            querying or used to derive information required for the migration.
        </li>
    </ul>
</p>
<p>

<li class="table_name">ident</li>
    <ul>
        <li>ident.feat_name</li>
        <li class="primary">ident.locus</li>
        <li class="primary">ident.com_name</li>
        <li class="primary">ident.comment</li>
        <li class="primary">ident.assignby</li>
        <li class="primary">ident.date</li>
        <li class="primary">ident.ec#</li>
        <li class="primary">ident.pub_comment</li>
        <li class="striked">ident.nt_comment</li>
        <li class="primary">ident.auto_comment</li>
        <li class="primary">ident.gene_sym</li>
        <li class="striked">ident.species</li>
        <li class="striked">ident.save_history</li>
        <li class="primary">ident.pub_locus</li>
        <li class="striked">ident.nt_com_name</li>
        <li class="primary">ident.is_pseudogene</li>
        <li class="primary">ident.alt_locus</li>
    </ul>

<li class="table_name">ident_xref</li>
    <ul>
        <li class="striked">ident_xref.id</li>
        <li>ident_xref.feat_name</li>
        <li class="striked">ident_xref.xref_id</li>
        <li class="primary">ident_xref.ident_val</li>
        <li class="primary">ident_xref.xref_type</li>
        <li class="striked">ident_xref.method</li>
        <li class="striked">ident_xref.mod_date</li>
        <li class="striked">ident_xref.assignby</li>
        <li class="primary">ident_xref.relrank</li>
    </ul>

<li class="table_name">ORF_attribute</li>
    <ul>
        <li class="striked">ORF_attribute.id</li>
        <li>ORF_attribute.feat_name</li>
        <li>ORF_attribute.att_type</li>
        <li class="primary">ORF_attribute.curated</li>
        <li class="striked">ORF_attribute.method</li>
        <li class="primary">ORF_attribute.date</li>
        <li class="primary">ORF_attribute.assignby</li>
	<li class="primary">score</li>
	<li class="primary">score2</li>
	<li class="primary">score_desc</li>
	<li class="primary">score3</li>
	<li class="primary">score_desc</li>
	<li class="primary">score4</li>
	<li class="primary">score4_desc</li>
	<li class="primary">score5</li>
	<li class="primary">score5_desc</li>
    </ul>

<li class="table_name">asm_feature</li>
    <ul>
        <li>asm_feature.feat_id</li>
        <li>asm_feature.feat_type</li>
        <li class="striked">asm_feature.feat_class</li>
        <li class="striked">asm_feature.feat_method</li>
        <li class="primary">asm_feature.end5</li>
        <li class="primary">asm_feature.end3</li>
        <li class="primary">asm_feature.comment</li>
        <li class="primary">asm_feature.assignby</li>
        <li class="primary">asm_feature.date</li>
        <li class="primary">asm_feature.sequence</li>
        <li class="primary">asm_feature.protein</li>
        <li class="primary">asm_feature.feat_name</li>
        <li class="striked">asm_feature.lock_id</li>
        <li>asm_feature.asmbl_id</li>
        <li class="striked">asm_feature.parent_id</li>
        <li class="striked">asm_feature.change_log</li>
        <li class="striked">asm_feature.is_gb</li>
        <li class="primary">asm_feature.pub_comment</li>
        <li class="striked">asm_feature.save_history</li>
        <li class="primary">asm_feature.curated</li>
    </ul>

<li class="table_name">sub_to_final</li>
    <ul>  
        <li>sub_to_final.asmbl_id</li>
        <li class="primary">sub_to_final.asm_lend</li>
        <li class="primary">sub_to_final.asm_rend</li>
        <li>sub_to_final.sub_asmbl_id</li>
        <li class="primary">sub_to_final.sub_asm_lend</li>
        <li class="primary">sub_to_final.sub_asm_rend</li>
    </ul>

<li class="table_name">evidence</li>
    <ul>  
        <li class="striked">evidence.id</li>
        <li>evidence.feat_name</li>
        <li>evidence.ev_type</li>
        <li class="primary">evidence.accession</li>
        <li class="primary">evidence.end5</li>
        <li class="primary">evidence.end3</li>
        <li class="primary">evidence.rel_end3</li>
        <li class="primary">evidence.rel_end5</li>
        <li class="primary">evidence.m_lend</li>
        <li class="primary">evidence.m_rend</li>
        <li class="secondary">evidence.curated</li>
        <li class="secondary">evidence.date</li>
        <li class="secondary">evidence.assignby</li>
        <li class="striked">evidence.change_log</li>
        <li class="striked">evidence.save_history</li>
        <li class="primary">evidence.method</li>
        <li class="primary">evidence.per_sim</li>
        <li class="primary">evidence.score</li>
        <li class="primary">evidence.db</li>
        <li class="primary">evidence.pvalue</li>
        <li class="primary">evidence.domain_score</li>
        <li class="primary">evidence.expect_domain</li>
        <li class="primary">evidence.total_score</li>
        <li class="primary">evidence.expect_whole</li>
        <li class="primary">evidence.chainID</li>
    </ul>

<li class="table_name">phys_ev</li>
<ul>
        <li>id</li>
	<li>feat_name</li>
	<li>ev_type</li>
	<li>assignby</li>
	<li>datestamp</li>
	<li>type</li>
	<li>score</li>
</ul>	

<li class="table_name">new_project</li>
<ul>
<li>id</li>
<li>notify</li>
<li>organism_name</li>
<li>1team_leader</li>
<li>project_leader</li>
<li>funded_by</li>
<li>grant#</li>
<li>genome_URL</li>
<li>taxon_id</li>
<li>genetic_code</li>
<li>lib_construct_date</li>
<li>seq_date</li>
<li>closure_date</li>
<li>annotation_date</li>
<li>lit_submit_date</li>
<li>publish_date</li>
<li>accession_submit_date</li>
<li>gram_stain</li>
<li>project_page</li>
<li>grant_deadline</li>
<li>asmbl_len</li>
<li>CMR_email</li>
<li>funding_URL</li>
<li>lineage</li>
<li>seq_db</li>
</ul>



<li class="table_name">assembly</li>
    <ul>  
        <li class="primary">assembly.asmbl_id</li>
        <li class="striked">assembly.seq_id</li>
        <li class="striked">assembly.bac_id</li>
        <li class="striked">assembly.com_name</li>
        <li class="striked">assembly.type</li>
        <li class="striked">assembly.method</li>
        <li class="striked">assembly.ed_status</li>
        <li class="striked">assembly.redundancy</li>
        <li class="striked">assembly.perc_N</li>
        <li class="striked">assembly.seq#</li>
        <li class="striked">assembly.full_cds</li>
        <li class="striked">assembly.cds_start</li>
        <li class="striked">assembly.cds_end</li>
        <li class="striked">assembly.ed_pn</li>
        <li class="striked">assembly.ed_date</li>
        <li class="striked">assembly.comment</li>
        <li class="primary">assembly.sequence</li>
        <li class="striked">assembly.frameshift</li>
        <li class="striked">assembly.release_date</li>
        <li class="striked">assembly.lock_id</li>
        <li class="striked">assembly.lsequence</li>
        <li class="striked">assembly.quality</li>
        <li class="striked">assembly.ca_contig_id</li>
        <li class="striked">assembly.mod_date</li>
        <li class="striked">assembly.is_circular</li>
    </ul>

<li class="table_name">go_role_link</li>
    <ul> 
        <li>go_role_link.id</li>
        <li>go_role_link.feat_name</li>
        <li class="primary">go_role_link.go_id</li>
        <li class="primary">go_role_link.assigned_by</li>
        <li class="primary">go_role_link.date</li>
        <li class="primary">go_role_link.qualifier</li>
        <li class="primary">go_role_link.nt_source</li>
    </ul>

<li class="table_name">go_evidence</li>
    <ul> 
        <li>go_evidence.role_link_id</li>
        <li class="primary">go_evidence.ev_code</li>
        <li class="primary">go_evidence.evidence</li>
        <li class="primary">go_evidence.with_ev</li>
    </ul>

<li class="table_name">gene_synonym</li>
    <ul>
      feat_name
      syn_feat_name
      assignby
      date
      synType
    </ul>

<li class="table_name">role_link</li>
    <ul>
      id
      parent_feat
      child_feat
      assignby
      datestamp
    </ul>


<li class="table_name">clone_info</li>
    <ul>
        clone_info.asmbl_id
	<li class="striked">clone_info.clone_id
	clone_info.clone_name
	clone_info.seq_group
	<li class="striked">clone_info.orig_annotation
	<li class="striked">clone_info.tigr_annotation
	clone_info.status
	clone_info.length
	clone_info.final_asmbl
	clone_info.fa_left
	clone_info.fa_right
	clone_info.fa_orient
	clone_info.gb_acc
	<li class="striked">clone_info.gb_desc
	<li class="striked">clone_info.gb_comment
	<li class="striked">clone_info.gb_date
	clone_info.comment
	clone_info.assignby
	clone_info.date
	clone_info.lib_id
	clone_info.seq_asmbl_id
	clone_info.chromo
	<li class="striked">clone_info.date_for_release
	<li class="striked">clone_info.date_released
	<li class="striked">clone_info.authors1
	<li class="striked">clone_info.authors2
	clone_info.seq_db
	clone_info.is_public
	<li class="striked">clone_info.gb_keywords
	clone_info.sequencing_type
	clone_info.prelim
	clone_info.license
	<li class="striked">clone_info.gb_gi
	<li class="striked">clone_info.gb_phase
   </ul>


	


<p class="section_header">
    Ignored tables
</p>
<p>
    This section lists tables that have been ignored during in the BSML migration.  This 
    information was parsed from the Sybase sysobjects table for a Prok database.
</p>
<p>
    <sql>select name from sysobjects where type = `U'</sql>
</p>
    Some of these tables are being ignored because it is not necessary to migrate the data 
    into BSML.  A large number of tables are being ignored because they are not used in 
    the annotation process.   Many of these ignored tables are used by the Sequencing/Closure 
    groups.
</p>
<p class="subsection_header">
    Ignored but used in annotation
</p>
     alignment<br>
     asm_feature_history<br>
     asmbl_link<br>
     asmbl_map<br>
     change_log<br>
     chromo_syn_link<br>
     evidence_history<br>
     ident_history<br>
<p>
</p>
<p class="subsection_header">
    Ignored and not used in annotation
</p>
<p>
    nt_ident<br>
    accession<br>
    annotator_lock<br>
    asmbl_tracker<br>
    bac_release<br>
    closure_status<br>
    edit_evaluation<br>
    edit_report<br>
    editor_sequence_lock<br>
    expert_subst_table<br>
    feature_history<br>
    frameshift<br>
    gb_authors<br>
    gb_citations<br>
    gb_feature<br>
    gb_release<br>
    marker<br>
    pull_worktable<br>
    region_evaluation<br>
    role_link<br>
    save_ed_lock<br>
    seq_contact<br>
    stan<br>
    subst_table<br>
</p>

<p class="section_header">
    Data types summary
</p>

<p class="subsection_header">
    Handled feature types
</p>
<p>
    The following asm_feature.feat_type values are documented in this migration document.
</p>
<p>
    CDS<br>
    TE<br>
    TU<br>
    acceptor<br>
    donor<br>
    exon<br>
    model<br>
    intergenic<br>
    pre-tRNA<br>
    rRNA<br>
    repeat<br>
    rna-exon<br>
    sRNA<br>
    slRNA<br>
    snRNA<br>
    snoRNA<br>
    tRNA<br>

</p>

<p class="subsection_header">
    Ignored feature types
</p>
<p>
    The following asm_feature.feat_type values from the asm_feature table are not being 
    handled during the BSML migration.
</p>
<p>
    6fseq
    misc_feature
    sfseq
    trf
    rscout
    subpiler
</p>

<p class="subsection_header">
    Handled ORF_attribute types
</p>
<p>
    The following ORF_attribute.att_type values were handled in the migration document.
</p>
<p>
    GC<br>
    GES<br>
    LP<br>
    MW<br>
    PI<br>
    SP<br>
    TERM<br>
    OMP<br>
    tRNA
</p>

<p class="subsection_header">
    Ignored ORF_attribute types
</p>
<p>
    The following ORF_attribute.att_type values from the ORF_attribute table are not being 
    handled during the BSML migration.
</p>
<p>
     MW
 PI
 GES
 OMP
 trf
 rRNA
 tRNA
 6fseg
 SP-NN
 sfseg
 slRNA
 snRNA
 SP-HMM
 psort2
 repeat
 snoRNA
 TEMPFAM
 targetP
 is_partial
 COG_curation
</p>

<p class="subsection_header">
    Handled evidence types
</p>
<p>
    BER<br>
    dds<br>
    nap<br>
    HMM2<br>
    gap2<br>
    para<br>
    BLASTP<br>
    NUCMER<br>
    autoGO<br>
    PROSITE<br>
    TBLASTN<br>
    Interpro<br>
    SEG_ngpd<br>
    FPrintScan<br>
    AUTO_BLASTP<br>
    BlastProDom<br>
    ProfileScan<br>
    ScanPrositeC<br>
    COG accession<br>
    alignAssembly<br>

</p>

<p class="subsection_header">
    Ignored evidence types
</p>

</body>

</html>





    










