<html><head><title>euk data types chart</title>
    
    <link rel="stylesheet" href="documentation.css">
    <style type="text/css">
        table.grid {
            border-collapse: collapse;
        }
        
        table.grid td {
            border: 1px solid #000000;
            padding: 2px 5px 2px 5px;
            text-align: center;
        }
        
        table.grid td.component {
            text-align: left;
        }
        
        table.grid td.header {
            background-color: #EEEEEE;
        }
        
        table.grid td.done {
            background-color: #82FF6E;
        }
        
        table.grid td.notdone {
            background-color: #FFFFFF;
        }

        table.grid td.na {
            background-color: #EEEEEE;
        }
	ul {
	    list-style-type: circle;
	}	
    </style></head>

<body>

<p class="doc_header">
    euk data types migration summary<br>
    <br>
    date created : 2005-09-12<br>
    last modified: 2005-09-14<br>
    <br>
    angiuoli@tigr.org<br>
    <br>
    This document provides a course overview of the current migration status of euk data types to
    Chado.  Detailed mapping information between the euk database and chado is here:
    <a href='http://tools/viewcvs/viewcvs.cgi/*checkout*/ANNOTATION/bsml/doc/legacy_eukdb_mapping_to_bsml.html?rev=HEAD&content-type=text/html'>[mapping_doc]</a>

</p>

<p class="section_header">
    euk data types migration summary
</p>
<p>
    <table class="grid">
        <tbody><tr>
            <td class="header">data class</td>
	    <td class="header">primary euk tables</td>
	    <td class="header">primary chado tables</td>
            <td class="header">mapping document</td>
            <td class="header">legacy2bsml</td>
            <td class="header">bsml2chado</td>
            <td class="header">validation script</td>
            <td class="header">test database</td>
            <td class="header">manual QC</td>
        </tr>
        <tr>
            <td class="component">genome info</td>
            <td class="na">new_project, common</td>
            <td class="na">organism, organism_prop</td>
            <td class="done"></td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">assembly seq/info</td>
            <td class="na">assembly, clone_info</td>
	    <td class="na">feature</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">genes</td>
            <td class="na">asm_feature, phys_ev, feat_link</td>
            <td class="na">feature, feature_relationship, featureloc</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">RNAs</td>
            <td class="na">asm_feature, phys_ev, feat_link</td>
            <td class="na">feature, feature_relationship, featureloc</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">Functional annotation</td>
            <td class="na">ident, go_role_link</td>
	    <td class="na">featureprop, feature_cvterm</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">ORF_attributes</td>
            <td class="na">ORF_attribute</td>
            <td class="na">featureprop, feature_cvterm</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">Scaffold mappings</td>
            <td class="na">sub_to_final, clone_info</td>
            <td class="na">featureloc</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">Gene finders</td>
            <td class="na">asm_feature, phys_ev</td>
            <td class="na">feature, featureloc</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
            <td class="done">&nbsp;</td>
            <td class="notdone">&nbsp;</td>
        </tr>
	<tr>
            <td class="component">Evidence</td>
            <td class="na">evidence</td>
            <td class="na">n/a</td>
            <td class="na">n/a</td>
            <td class="na">n/a</td>
            <td class="na">n/a</td>
            <td class="na">n/a</td>
            <td class="na">n/a</td>
            <td class="na">n/a</td>
        </tr>
</table>
</p>

<p class="section_header">
Genome info
</p>
<ul>
<li>Organism name</li>
</ul>
<p class="section_header">
Assembly seq/info
<ul>
<li>Sequence feature</li>
<li>clone_info attributes</li>
<li>SO type: contig, supercontig, ...</li>
</ul>
<p class="section_header">
Genes
<ul>
<li>Features
<ul>
<li>TU
<li>working model
<li>CDS
<li>exons
</ul>
<li>Sequences
<ul>
<li>CDS sequence
<li>protein translation
</ul>
</ul>
</p>
<p class="section_header">
RNAs
<ul>
<li>Features
<ul>
<li>ncRNA
<li>pre-tRNA
<li>rna-exon
<li>tRNA
</ul>
</ul>
</p>
<p class="section_header">
Functional annotation
<ul>
<li>"Ident" info
<ul>
<li>gene_product_name
<li>gene_name
<li>ec_number
<li>gene_symbol
<li>assignby,date
<li>comment
</ul>
<li>GO annotation + evidence
</ul>
<p class="section_header">
"ORF_attributes"
<ul>
<li>MW
<li>PI
</ul>
</p>

<p class="section_header">
Gene finders
<ul>
<li>Any model != 'working' is migrated </li>
<li>Support for user specified list </li>
</ul>
</p>





</body></html>
